PluginsAndFeatures/azure-toolkit-for-intellij/azure-intellij-plugin-azure-sdk/src/main/java/com/microsoft/azure/toolkit/intellij/azure/sdk/buildtool/UseOfBlockOnAsyncClientsCheck.java

```diff
@@ -0,0 +1,165 @@
+        public void visitMethodCallExpression(@NotNull PsiMethodCallExpression expression) {
+            super.visitMethodCallExpression(expression);
+
+            if (SKIP_WHOLE_RULE) {
+                return;
+            }
+
+            // Check if the method call is block() or a variant of block() method
+            if (expression.getMethodExpression().getReferenceName().startsWith(RULE_CONFIG.getMethodsToCheck().get(0))) {
+
+                // Check if the method call is on a reactive type
+                boolean isAsyncContext = checkIfAsyncContext(expression);
+
+                if (isAsyncContext) {
+                    holder.registerProblem(expression, RULE_CONFIG.getAntiPatternMessageMap().get("antiPatternMessage"));
+                }
+            }
+        }
+
+        /**
+         * This method is used to check if the method call is a block() method call on a reactive type.
+         * The method call is checked to see if it is a block() method call on a reactive type
+         * and if the reactive type is an async client in Azure SDK.
+         *
+         * @param methodCall PsiMethodCallExpression - the method call expression to check
+         * @return true if the method call is a block() method call on an async client, false otherwise
+         */
+        private boolean checkIfAsyncContext(@NotNull PsiMethodCallExpression methodCall) {
+
+            // Get the qualifier expression of the method call -- the expression before the 'block' method call
+            PsiExpression qualifierExpression = methodCall.getMethodExpression().getQualifierExpression();
+
+            if (qualifierExpression instanceof PsiMethodCallExpression) {
+                PsiMethodCallExpression qualifierMethodCall = (PsiMethodCallExpression) qualifierExpression;
+
+                // Get the return type of the qualifier method call
+                PsiType qualifierReturnType = qualifierMethodCall.getType();
+
+                if (qualifierReturnType instanceof PsiClassType) {
+                    PsiClass qualifierReturnTypeClass = ((PsiClassType) qualifierReturnType).resolve();
+
+                    // Check if the return type is a subclass of Mono or Flux
+                    if (qualifierReturnTypeClass != null && isReactiveType(qualifierReturnTypeClass)) {
+
+                        return isAzureAsyncClient(qualifierMethodCall);
+                    }
+                }
+            }
+            return false;
+        }
+
+        /**
+         * This method is used to check if the method call is on an async client in Azure SDK.
+         * The method call is checked to see if it is on an async client in Azure SDK.
+         *
+         * @param qualifierMethodCall PsiMethodCallExpression - the method call expression to check
+         * @return true if the method call is on an async client in Azure SDK, false otherwise
+         */
+        private boolean isAzureAsyncClient(PsiMethodCallExpression qualifierMethodCall) {
```

Comment:
Do we want to update this rule to check for block on async operations on client methods rather than on async clients.

Reply:
I check that block is on an async operation first then check for async azure clients so its just an additional step to make it azure-specific